File: frontend/src/types/safety/central-alarm.ts
// ============ 中控报警分析（central-alarm-analysis）Types ============
// 契约对齐：ticket 03（records/stats，分布为 dict）、ticket 05/06（日常报告 report_kind/analyzed）

/** AI 异常模式（英文枚举，后端归一化后回写；前端经 AI_PATTERN_CONFIG 映射中文） */
export type CentralAlarmPattern =
  | 'normal_transient'
  | 'repeated'
  | 'false_alarm'
  | 'anomalous'

/** AI 分析维度（英文枚举，后端归一化后回写） */
export type CentralAlarmDimension = 'process' | 'operation' | 'equipment' | 'other'

/** 报警记录（列表/详情共用，来源 safety.central_alarm_records） */
export interface CentralAlarmRecord {
  id: string
  feishu_record_id?: string | null
  source?: string | null
  alarm_date?: string | null // ISO 时间，前端 dayjs 格式化
  post?: string | null // 岗位（Bitable 单选）
  alarm_description?: string | null // 报警情况说明（自由文本）
  special_note?: string | null // 特殊情况说明
  workshop?: string | null // 车间（如 车间一）
  line?: string | null // 产线/区域（如 达托）
  // AI 分析字段组（日报生成时回写）
  ai_alarm_type?: string | null // 结构化报警类型（高液位/高温/误报等）
  ai_equipment?: string | null // AI 抽取设备
  ai_pattern?: string | null // 取值见 CentralAlarmPattern（未知值前端降级灰 Tag 原文显示）
  ai_dimension?: string | null
  ai_reason_analysis?: string | null
  ai_rectification_direction?: string | null
  ai_analyzed_at?: string | null
  synced_at?: string | null
  created_at?: string | null
}

/** 列表查询参数（日期 YYYY-MM-DD） */
export interface CentralAlarmQueryParams {
  page?: number
  page_size?: number
  date_from?: string
  date_to?: string
  workshop?: string
  line?: string
  post?: string
  ai_alarm_type?: string
  ai_dimension?: string
  ai_pattern?: string
  keyword?: string
}

/** KPI 统计（分布均为 dict[str, int]，统计口径默认本周） */
export interface CentralAlarmStats {
  date: string
  today_total: number
  week_total: number
  total_count: number
  week_ai_analyzed_count: number
  workshop_distribution: Record<string, number>
  post_distribution: Record<string, number>
  alarm_type_distribution: Record<string, number>
  pattern_distribution: Record<string, number>
  dimension_distribution: Record<string, number>
}

/** Bitable 多表全量同步结果 */
export interface CentralAlarmSyncResult {
  synced_count: number
  soft_deleted_count?: number
}

/** 日报生成请求（target_date 可选，默认当天·北京时区） */
export interface CentralAlarmDailyReportRequest {
  target_date?: string
}

/** 推送结果项 */
export interface CentralAlarmPushResult {
  chat_id: string
  success: boolean
  message_id?: string
  skipped?: boolean
  reason?: string
  error?: string
}

/** 日报生成响应（不落库，前端弹窗展示） */
export interface CentralAlarmReportResponse {
  report_kind: 'daily'
  target_date: string
  total: number
  analyzed: number
  markdown_report: string
  push_results: CentralAlarmPushResult[]
  records_analyzed?: string[]
}

File: backend/app/modules/safety/schemas/central_alarm.py
"""中控报警 Schema — API 契约（ticket 02 同步；ticket 03 查询/统计；ticket 05/06 日报生成）。"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


class CentralAlarmSyncResponse(BaseModel):
    """Bitable 多表全量同步结果。"""

    synced_count: int = Field(..., description="本次同步 upsert 的记录数")
    soft_deleted_count: int = Field(0, description="本次同步软删除的记录数")


class CentralAlarmRecordOut(BaseModel):
    """报警记录列表项（含 AI 分析列）。"""

    id: uuid.UUID
    feishu_record_id: str | None = None
    source: str
    alarm_date: datetime | None = None
    post: str | None = None
    alarm_description: str | None = None
    special_note: str | None = None
    workshop: str | None = None
    line: str | None = None
    ai_alarm_type: str | None = None
    ai_equipment: str | None = None
    ai_pattern: str | None = None
    ai_dimension: str | None = None
    ai_reason_analysis: str | None = None
    ai_rectification_direction: str | None = None
    ai_analyzed_at: datetime | None = None
    synced_at: datetime | None = None
    created_at: datetime | None = None


class CentralAlarmStatsOut(BaseModel):
    """KPI 卡数据（分布为 dict 而非数组）。"""

    date: str  # 当日日期 ISO（target_date 或今天·北京时间）
    today_total: int = 0  # 今日报警数
    week_total: int = 0  # 本周（自然周 周一~周日·北京时间）报警数
    total_count: int = 0  # 全量报警数（非软删）
    week_ai_analyzed_count: int = 0  # 本周已 AI 分析的记录数
    workshop_distribution: dict[str, int] = Field(default_factory=dict)  # 车间 → 数量
    post_distribution: dict[str, int] = Field(default_factory=dict)  # 岗位 → 数量
    alarm_type_distribution: dict[str, int] = Field(default_factory=dict)  # 报警类型 → 数量
    pattern_distribution: dict[str, int] = Field(default_factory=dict)  # 异常模式 → 数量
    dimension_distribution: dict[str, int] = Field(default_factory=dict)  # AI 维度 → 数量


class CentralAlarmDailyReportRequest(BaseModel):
    """日报生成请求（ticket 06）。"""

    target_date: date | None = Field(None, description="目标日期，默认今天（北京时区）")


class CentralAlarmReportResponse(BaseModel):
    """日报生成结果（不落库，前端弹窗展示）。"""

    report_kind: str = "daily"
    target_date: date
    total: int = 0
    analyzed: int = 0
    markdown_report: str
    push_results: list[dict] = Field(default_factory=list)
    records_analyzed: list[uuid.UUID] = Field(default_factory=list)
